lib/rpush/daemon/apns2/delivery.rb

```diff
@@ -0,0 +1,132 @@
+module Rpush
+  module Daemon
+    module Apns2
+      # https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/CommunicatingwithAPNs.html
+      class Delivery < Rpush::Daemon::Delivery
+        RETRYABLE_CODES = [ 429, 500, 503 ]
+
+        def initialize(app, http2_client, batch, notification)
+          @app = app
+          @client = http2_client
+          @batch = batch
+          @notification = notification
+        end
+
+        def perform
+          handle_response(do_post)
+        rescue SocketError => error
+          mark_retryable(@notification, Time.now + 10.seconds, error)
+          raise
+        rescue StandardError => error
+          mark_failed(error)
+          reflect(:error, error)
+          raise
+        ensure
+          @batch.notification_processed
+        end
+
+        protected
+        ######################################################################
+
+        def handle_response(response)
+          code = response.code
+          case code
+          when 200
+            ok(response)
+          when *RETRYABLE_CODES
+            service_unavailable(response)
+          else
+            reflect(:notification_id_failed,
+              @app,
+              @notification.id, code,
+              response.failure_reason)
+            fail Rpush::DeliveryError.new(response.code,
+              @notification.id, response.failure_reason)
+          end
+        end
+
+        def ok(response)
+          log_info("#{@notification.id} sent to #{@notification.device_token}")
+          mark_delivered
+        end
+
+        def service_unavailable(response)
+          mark_retryable(@notification, Time.now + 10.seconds)
+          # Logs should go last as soon as we need to initialize
+          # retry time to display it in log
+          failed_message_to_log(response)
+          retry_message_to_log
+        end
+
+        def do_post
+          request = prepare_request
+          resp = @client.call(:post, request[:path],
```

Comment:
Not sure if we win something with streaming requests in this particular case, the case is very simple and request/response is pretty short, but for the future it makes sense. Check if everything is fine please, and thanks for review!